src/go/MONIT/alert.go

```diff
@@ -0,0 +1,658 @@
+func (e endsAtSorter) Less(i, j int) bool {
+	return e[j].EndsAt.After(e[i].EndsAt)
+}
+
+//Function for sorting alerts based on a passed label
+func sortAlert() {
+	severityLevel = make(map[string]int)
+	//TODO : Have to add all possibilities of severity levels.
+	severityLevel["info"] = 0
+	severityLevel["warning"] = 1
+	severityLevel["medium"] = 2
+	severityLevel["high"] = 3
+	severityLevel["critical"] = 4
+	severityLevel["urgent"] = 5
+
+	switch strings.ToLower(sortLabel) {
+	case "severity":
+		sort.Sort(severitySorter(allAlertData))
+	case "starts":
+		sort.Sort(startAtSorter(allAlertData))
+	case "ends":
+		sort.Sort(endsAtSorter(allAlertData))
+	case "duration":
+		sort.Sort(durationSorter(allAlertData))
+	default:
+		return
+	}
+}
+
+//Function for printing alerts in JSON format
+func jsonPrint() {
+
+	var filteredData []alertDataJSON
+	var temp alertDataJSON
+
+	for _, each := range allAlertData {
+		if filteredAlerts[each.Name] == 1 {
+			temp.Name = each.Name
+			temp.Service = each.Service
+			temp.Severity = each.Service
+			temp.Tag = each.Tag
+			temp.Starts = timeDiff(time.Now(), each.StartsAt, 0)
+			if each.EndsAt == time.Unix(maxtstmp, 0).UTC() {
+				temp.Ends = "Undefined"
+				temp.Duration = "Undefined"
+			} else {
+				temp.Ends = timeDiff(time.Now(), each.EndsAt, 0)
+				temp.Duration = timeDiff(each.StartsAt, each.EndsAt, 1)
+			}
+
+			filteredData = append(filteredData, temp)
+		}
+	}
+
+	b, err := json.Marshal(filteredData)
+
+	if err != nil {
+		log.Printf("Unable to convert Filtered JSON Data, error: %v\n", err)
+		return
+	}
+
+	fmt.Println(string(b))
+}
+
+//Function for printing alerts in Plain text format
+func tabulate() {
+
+	w := new(tabwriter.Writer)
+	w.Init(os.Stdout, 8, 8, 0, '\t', 0)
+	defer w.Flush()
+
+	fmt.Fprintf(w, "\n %s\t\t%s\t\t%s\t\t%s\t\t%s\t\t%s\t\t%s\n", "NAME", "SERVICE", "TAG", "SEVERITY", "STARTS", "ENDS", "DURATION")
+
+	for _, each := range allAlertData {
+		if filteredAlerts[each.Name] == 1 {
+			fmt.Fprintf(w, " %s\t\t%s\t\t%s\t\t%s\t\t%s",
+				each.Name,
+				each.Service,
+				each.Tag,
+				each.Severity,
+				timeDiff(time.Now(), each.StartsAt, 0),
+			)
+			if each.EndsAt == time.Unix(maxtstmp, 0).UTC() {
+				fmt.Fprintf(w, "\t\t%s", "Undefined")
+				fmt.Fprintf(w, "\t\t%s\n", "Undefined")
+			} else {
+				fmt.Fprintf(w, "\t\t%s", timeDiff(time.Now(), each.EndsAt, 0))
+				fmt.Fprintf(w, "\t\t%s\n", timeDiff(each.StartsAt, each.EndsAt, 1))
+			}
+
+		}
+	}
+}
+
+//Function for printing alert's details in JSON format
+func jsonPrintDetails() {
+
+	b, err := json.Marshal(alertDetails[name])
+
+	if err != nil {
+		log.Printf("Unable to convert Detailed JSON Data, error: %v\n", err)
+		return
+	}
+
+	fmt.Println(string(b))
+}
+
+//Function for printing alert's details in Plain text format
+func detailPrint() {
+
+	if each, ok := alertDetails[name].(amGGUSJSON); ok {
+		fmt.Printf("NAME: %s\n", name)
+		fmt.Printf("LABELS\n")
+		fmt.Printf("\tservice: %s\n", each.Labels.Service)
```

Comment:
here again everything will be generalized if you will use Labels/Annotations as map[string]interace{}. This will allow you to write two loops: one over Labels and another over annotations. The code will only have one print statement, e.g.
```
for k, v := range each.Labels {
  fmt.Printf("\t%s: %v\n", k, v)
}
for k, v := range each.Annotations {
  fmt.Printf("\t%s: %v\n", k, v)
}
```

Reply:
Valentin,
Thanks, the loop over the maps are very helpful. Implemented the same.
However, I had to add two loops for LABELS because of the use of map[string]interface{} in amJSON which in turn causes unordered LABELS. 
```
if alert, ok := alertDetails[name]; ok { //use of two different loops for LABELS as sometimes AM was sending unordered LABELS i.e.
for key, value := range alert.Labels { //"severity", "tag", "service" could arrive before "alertname"
if key == "alertname" {
	fmt.Printf("NAME: %s\n", value)
	break
}
}
fmt.Printf("LABELS\n")
for key, value := range alert.Labels {
switch key {
case "service", "tag", "severity":
	fmt.Printf("\t%s: %s\n", key, value)
}
}
```

Reply:
Why not to get list of map keys and sort them out. Then loop over sorted keys and perform one loop? I use this all the time, e.g.
https://github.com/dmwm/das2go/blob/master/utils/utils.go#L111

Reply:
Fixed this issue.